storage/src/interface.rs

```diff
@@ -0,0 +1,190 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! High-level application-agnostic storage interface
+
+use crate::schema::{self, Schema};
+use storage_core::{backend, Backend, Data, DbIndex};
+
+/// The main storage type
+pub struct Storage<B: Backend, Sch> {
+    backend: B::Impl,
+    _schema: core::marker::PhantomData<Sch>,
+}
+
+impl<B: Backend, Sch> Clone for Storage<B, Sch>
+where
+    B::Impl: Clone,
+{
+    fn clone(&self) -> Self {
```

Comment:
I think a derive would work for this. No need to specialize the clone() implementation.

Reply:
I just checked and it does not work because `self` contains a `B::Impl` and not `B`. The auto-derived `Clone` impl stupidly puts in the `B: Clone` constraint whereas `B::Impl: Clone` is needed.